### PR TITLE
Store RT, add scopes to static-spa sample

### PR DESCRIPTION
**Pull request by swiftone**

### CODE REVIEW

File: lib/token.ts [hunk 1]
```diff
@@ -272,6 +273,16 @@ function handleOAuthResponse(sdk: OktaAuth, tokenParams: TokenParams, res: OAuth
       };
     }
 
+    if (refreshToken) {
+      tokenDict.refreshToken = {
+        refreshToken: refreshToken,
+        value: refreshToken,
```
Comment (aarongranick-okta) [inline lib/token.ts:279]: let's avoid "value". We have an existing task to remove this. https://github.com/okta/okta-auth-js/pull/515

File: samples/generated/static-spa/public/app.js [hunk 2]
```diff
@@ -266,7 +269,8 @@ function redirectToGetTokens(additionalParams) {
 
 function redirectToLogin(additionalParams) {
   authClient.token.getWithRedirect(Object.assign({
-    state: JSON.stringify(config.state)
+    state: JSON.stringify(config.state),
+    scopes: config.state.scopes.split(/\s+/) || config.scopes, // getWithRedirect doesn't obey scopes in constructor yet
```
Comment (aarongranick-okta) [inline samples/generated/static-spa/public/app.js:273]: is this a bug?

File: lib/token.ts (hunk 1 above)
Comment (swiftone) [reply lib/token.ts:279]: I started that and saw that it'd be changing the base type of Token so I held off, but I can give it another go.

File: samples/generated/static-spa/public/app.js (hunk 2 above)
Comment (swiftone) [reply samples/generated/static-spa/public/app.js:273]: I certainly think so.

File: lib/token.ts (hunk 1 above)
Comment (shuowu) [reply lib/token.ts:279]: #515 is a breaking change since type would be changed. btw, can anyone approve that pr.

File: lib/TokenManager.ts [hunk 3]
```diff
@@ -153,7 +154,7 @@ function validateToken(token: Token) {
   if (!isObject(token) ||
       !token.scopes ||
       (!token.expiresAt && token.expiresAt !== 0) ||
-      (!isIDToken(token) && !isAccessToken(token))) {
+      (!isIDToken(token) && !isAccessToken(token) && !isRefreshToken(token))) {
     throw new AuthSdkError('Token must be an Object with scopes, expiresAt, and an idToken or accessToken properties');
```
Comment (vijetmahabaleshwar-okta) [inline lib/TokenManager.ts:158]: nit: Should the error say "and idToken, accessToken or refreshToken properties"?

File: lib/token.ts [hunk 4]
```diff
@@ -724,14 +736,59 @@ function renewToken(sdk: OktaAuth, token: Token): Promise<Token> {
...
-    return isIDToken(token) ? tokens.idToken : tokens.accessToken;
-  });
+    .then(function (res) {
+      // Multiple tokens may have come back. Return only the token which was requested.
+      var tokens = res.tokens;
+      return isIDToken(token) ? tokens.idToken : tokens.accessToken;
+    });
+}
+
+async function renewTokensWithRefresh(sdk: OktaAuth, options: TokenParams, refreshTokenObject: RefreshToken): Promise<any> {
```
Comment (shuowu) [inline lib/token.ts:746]: Looks like the return type is TokenResponse instead of Tokens, which means the tokenManager cannot get proper tokens after the renew process.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,20 @@
 # Changelog
 
+## PENDING
+
+### Features
+- Adding the ability to use refresh tokens with single page applications (SPA) (Early Access feature - reach out to our support team)
+  - `scopes` configuration option now handles 'offline_access' as an option, which will use refresh tokens IF the app of the clientId being used is configured to do so in the Okta settings
```
Comment (vijetmahabaleshwar-okta) [inline CHANGELOG.md:7]: nit: "if the app of the clientId being used" sounds a little off. Can we just tell "If your client app is configured"? Unless we specifically want to mention `clientId` I guess...

Comment (swiftone) [reply CHANGELOG.md:7]: I was trying to make sure they understood the configuration wasn't in their code (where the `scopes` value under discussion is).  Agree it is awkward though - open to anything that makes that distinction clear.

File: lib/TokenManager.ts [hunk 6]
```diff
@@ -304,10 +316,14 @@ function renew(sdk, tokenMgmtRef, storage, key) {
...
+  // A refresh token means a refresh instead of renewal
+  // TODO: XXX  
+
   // Store the renew promise state, to avoid renewing again
-  // Renew both tokens in one process
+  // Renew/refresh all tokens in one process
   tokenMgmtRef.renewPromise[key] = sdk.token.renewTokens({
-    scopes: token.scopes
+    scopes: token.scopes,
+    refreshToken: storage.getStorage()
```
Comment (aarongranick-okta) [inline lib/TokenManager.ts:326]: doublecheck this

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,19 @@
 # Changelog
 
+## PENDING
```
Comment (aarongranick-okta) [inline CHANGELOG.md:3]: Can make this 4.2.0. package.json is already bumped, will be a normal release from master

File: samples/generated/static-spa/public/app.js [hunk 8]
```diff
@@ -335,6 +354,7 @@ function showForm() {
   // Set values from config
   document.getElementById('issuer').value = config.issuer;
   document.getElementById('clientId').value = config.clientId;
+  document.getElementById('scopes').value = config.scopes;
```
Comment (aarongranick-okta) [inline samples/generated/static-spa/public/app.js:357]: I think it would be less confusing if `config.scopes` in the samples was an array, consistent with what okta-auth-js expects. The conversion to/from string can be handled here in `showForm`. In the default test app, the array is converted to a comma separated string: https://github.com/okta/okta-auth-js/blob/master/test/app/src/form.ts#L51
Within the JSON serialized state param, it can remain an array. It only needs to be a string for the 2 cases:
- in the form
- as a standalone query parameter
These 2 cases are closely related because the form has method GET so submitting the form will set the query parameter with the value in the form.  The serialized state object is used to save the config across redirects.

**Comment by aarongranick-okta** [comment]
I tested it and it works. Nice job!

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,19 @@
...
+
+### Features
+- Adding the ability to use refresh tokens with single page applications (SPA) (Early Access feature - reach out to our support team)
+  - `scopes` configuration option now handles 'offline_access' as an option, which will use refresh tokens IF your client app is configured to do so in the Okta settings
+    - If you already have tokens (from a separate instance of auth-js or the okta-signin-widget) those tokens must already include a refresh token and have the 'offline_access' scope
+    - 'offline_access' is not requested by default.  Anyone using the default `scopes` and wishing to add 'offline_access' should pass `scopes: ['openid', 'email', 'offline_access']` to their constructor (FIXME: Is constructor enough?)
```
Comment (shuowu) [inline CHANGELOG.md:9]: Aaron has a fix for 4.1, should be backported after merging https://github.com/okta/okta-auth-js/pull/535

File: lib/AuthStateManager.ts [hunk 10]
```diff
@@ -136,7 +137,7 @@ export class AuthStateManager {
       };
 
       this._sdk.tokenManager.getTokens()
-        .then(({ accessToken, idToken }) => {
+        .then(({ accessToken, idToken, refreshToken }) => {
```
Comment (shuowu) [inline lib/AuthStateManager.ts:140]: As discussed, should we hide the refreshToken from authState?

File: lib/browser/browser.ts [hunk 11]
```diff
@@ -333,17 +350,26 @@ class OktaAuthBrowser extends OktaAuthBase implements OktaAuth, SignoutAPI {
...
+    if (revokeRefreshToken && typeof refreshToken === 'undefined') {
+      refreshToken = (await this.tokenManager.getTokens()).refreshToken as RefreshToken;
+    }
+
     if (revokeAccessToken && typeof accessToken === 'undefined') {
       accessToken = (await this.tokenManager.getTokens()).accessToken as AccessToken;
     }
   
     // Clear all local tokens
     this.tokenManager.clear();
```
Comment (shuowu) [inline lib/browser/browser.ts:363]: We may want to emit token change events when clear all tokens to update authState properly. Can be done in another PR.

File: lib/browser/browser.ts [hunk 12]
```diff
@@ -427,6 +453,11 @@ class OktaAuthBrowser extends OktaAuthBase implements OktaAuth, SignoutAPI {
     return accessToken ? accessToken.accessToken : undefined;
   }
 
+  getRefreshToken(): string | undefined {
```
Comment (shuowu) [inline lib/browser/browser.ts:456]: remove this method if we want to expose refreshToken only internally for now.

File: CHANGELOG.md (hunk 7 above)
Comment (swiftone) [reply CHANGELOG.md:3]: My inclination is to leave this as PENDING until we're ready for release, as it helps with people on github asking why the version on NPM doesn't agree with what they are seeing, but I'm open to persuasion.

File: lib/AuthStateManager.ts (hunk 10 above)
Comment (swiftone) [reply lib/AuthStateManager.ts:140]: I'm not sure we should entirely HIDE it - so much as let people not have to worry about it.

File: lib/token.ts (hunk 1 above)
Comment (swiftone) [reply lib/token.ts]: updating this comment for posterity: Decision was to leave .value to keep this from being a breaking change, but we'll remove it from everything in 5.0

File: lib/browser/browser.ts [hunk 13]
```diff
@@ -311,7 +312,21 @@ class OktaAuthBrowser extends OktaAuthBase implements OktaAuth, SignoutAPI {
     return this.token.revoke(accessToken);
   }
 
-  // Revokes accessToken, clears all local tokens, then redirects to Okta to end the SSO session.
+  // Revokes the refresh token for the application session
+  async revokeRefreshToken(refreshToken?: RefreshToken) {
+    if (!refreshToken) {
+      refreshToken = (await this.tokenManager.getTokens()).refreshToken as RefreshToken;
```
Comment (shuowu) [inline lib/browser/browser.ts:318]: It might be better to get the token from authState (sync access), but as we are also using the same logic in `revokeAccessToken`, the change can be made in a separate PR.

File: lib/browser/browser.ts [hunk 14]
```diff
@@ -426,6 +451,11 @@ class OktaAuthBrowser extends OktaAuthBase implements OktaAuth, SignoutAPI {
     return accessToken ? accessToken.accessToken : undefined;
   }
 
+  getRefreshToken(): string | undefined {
```
Comment (shuowu) [inline lib/browser/browser.ts:454]: so we decided to keep the public method for refresh token?

Comment (shuowu) [reply lib/browser/browser.ts:454]: looks like it's neither in readme nor used in code, remove?

**Review by shuowu — APPROVED** [review]

File: lib/browser/browser.ts (hunk 13 above)
Comment (swiftone) [reply lib/browser/browser.ts:318]: We get into a circular dependency if we try that.   I brought it up and Aaron suggested we just do this.

File: lib/browser/browser.ts (hunk 14 above)
Comment (swiftone) [reply lib/browser/browser.ts:454]: We kept it because we're exposing it on authState anyway.

File: samples/config.js [hunk 15]
```diff
@@ -20,6 +20,7 @@ const defaults = {
 const spaDefaults = Object.assign({
   redirectPath: '/implicit/callback',
   flow: 'redirect',
+  scopes: 'openid email',
```
Comment (aarongranick-okta) [inline samples/config.js:23]: I'm still not a fan of having scopes as a space-separated string here in the config. I would much prefer that scopes is an array, as it is for OktaAuth, and the conversion to string (comma or space separated) was done in the showForm() and conversion from string back to array is in loadConfig()

**Review by aarongranick-okta — APPROVED** [review]
see comments about "scopes" in the samples

File: lib/browser/browser.ts (hunk 13 above)
Comment (shuowu) [reply lib/browser/browser.ts:318]: wondering how the circular dep happens? it should just be an access to the in-memory state

```
const { refreshToken } = this.authStateManager.getAuthState();
```

Comment (swiftone) [reply lib/browser/browser.ts:318]: I'm going off memory, but as I recall `this.authStateManager` didn't exist here.
